### DISCUSSION TITLE
dp_ipi error

### DISCUSSION THREAD

**Discussion by neojie** [deepmodeling/deepmd-kit#308] [Q&A]
I was trying to run the water example of dp_ipi given in the official tutorial,
```
{
    "verbose":		false,
    "use_unix":		true,
    "port":		31415,
    "host":		"localhost",
    "graph_file":	"graph.pb",
    "coord_file":	"conf.xyz",
    "atom_type" : {
	"OW":		0, 
	"HW1":		1,
	"HW2":		1
    }
}
```
 but the following error
```
Error opening socket: wrong host address, or broken connection: No such file or directory
```

I was running in a cluster node.

**Comment by y1xiaoc** [comment]
Have you started the i-pi server first? (by running the i-pi command)
When setting `"use_unix": true`, dp_ipi would try to look up a local unix socket (which is a file like object) in your file system. It looks  like it failed to find the object. Running i-pi server first should create that socket object for you. 
If it still fails, please provide the detailed commands you execute and the xml i-pi config file so we can help to find what's wrong.
Thanks!

**Comment by neojie** [comment]
I did try to i-pi input first. What I did is
`i-pi input.xml > log &`
`cat input.xml` gives
```
<simulation verbosity='high'>
  <output prefix='simulation'>
    <properties stride='1' filename='out'>  [ step, time{picosecond}, conserved{electronvolt}, temperature{kelvin}, kinetic_cv{electronvolt}, potential{electronvolt}, pressure_cv{megapascal} ] </properties>
    <trajectory filename='pos' stride='20'> positions </trajectory>
    <checkpoint stride='200'/>
  </output>
  <total_steps>10000</total_steps>
  <prng>
    <seed>31415</seed>
  </prng>
  <ffsocket name='dp_ipi' mode='unix'>
    <address>example</address>
  </ffsocket>
  <system>
    <initialize nbeads='1'>
      <file mode='pdb'> init.pdb </file>
      <velocities mode='thermal' units='kelvin'> 300 </velocities>
    </initialize>
    <forces>
      <force forcefield='dp_ipi'> </force>
    </forces>
    <motion mode='dynamics'>
      <dynamics mode='nvt'>
        <timestep units='femtosecond'> 0.5 </timestep>
        <thermostat mode='svr'>
          <tau units='femtosecond'> 100 </tau>
        </thermostat>
      </dynamics>
    </motion>
    <ensemble>
      <temperature units='kelvin'> 300 </temperature>
    </ensemble>
  </system>
</simulation>
```
Note I really do not know what I should input for `ffsocket name`. So I guess it `dp_ipi`. Same for `force forcefield`.
After running `i-pi input.xml > log &`, I got files including `log simulation.out  simulation.pos_0.xyz  simulation.restart`.
Then I ran `dp water.json` in the same terminal.
Here `water.json` is

```
{
    "verbose":          true,
    "use_unix":         true,
    "port":             31415,
    "host":             "localhost",
    "graph_file":       "graph.pb",
    "coord_file":       "init.pdb",
    "atom_type" : {
        "OW":           0,
        "HW1":          1,
        "HW2":          1
    }
}

But again, I got error `Error opening socket: wrong host address, or broken connection: No such file or directory`

```

**Comment by y1xiaoc** [comment]
Try changing the `<address>example</address>` in `ffsocket` to `<address>localhost</address>`. It should match with the one in the `"host"` key of `water.json`. It doesn't have to be localhost though since you are using unix socket. Any legal file name should be possible.

**Comment by neojie** [comment]
Cool. It runs now. Two questions I still have:
1- What is the format of `atom_type` in the `water.json` file? For example, `OW` means `Oxygen`, but what is this `W` here?  Sam for `"HW1 and HW2". `I guess `0` in `"OW":           0,` means type `0`, which should be in line with the `type_map.raw` of the training set. 
2-Any other flags that I can set for this water.json file? Or that is all?

**Comment by y1xiaoc** [comment]
The names like "OW" "HW1" correspond to the atom name you provided in the conf.xyz. In normal xyz it may be simply "O" and "H". Here the example is taking from some classical force field I guess so it has to label the type of different oxygens and hydrogens. For example, the example `conf.xyz` may look like the following (the coordinates can all be zero)
```
6
some comments
OW 0 0 0
HW1 0 0 0
HW2 0 0 0
OW 0 0 0
HW1 0 0 0
HW2 0 0 0
```
And the `0` is correspond to the type in your training set indeed. Please refer to the last paragraph of the [tutorial](https://github.com/deepmodeling/deepmd-kit#run-path-integral-md-with-i-pi) .

I don't think there are more flags available. Most configurations should be set with ipi's xml file. You may refer to the [source file](https://github.com/deepmodeling/deepmd-kit/blob/master/source/ipi/driver.cc) for detailed implementation.

**Comment by neojie** [comment]
I see. I notice that the last paragraph of the tutorial says `The dp_ipi gets the atom names from an XYZ file provided by coord_file (meanwhile ignores all coordinates in it), and translates the names to atom types by rules provided by atom_type`. 
Also, the source codes only contain `XyzFileManager.h`. I guess this mean that only `.xyz` structure file will be read by dp_ipi. Therefore, I should construct a `.xyz` file and revise my water.json to `"coord_file":       "init.xyz",`. It is weird that I used `"coord_file":       "init.pdb",` but still get the simulation run. It seems that dp_ipi has handled this exception but I cannot find the details in the source file.
